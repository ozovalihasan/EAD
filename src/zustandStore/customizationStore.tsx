import { availableColors, availableColorsType, setColorVariants } from '@/helpers';
import create from 'zustand';
import { devtools } from 'zustand/middleware'


export enum LocationSidebar {
  Left = "left",
  Right = "right"
}

export interface CustomizationStoreState {
  locationSidebar: LocationSidebar;
  widthSidebar: number;
  showTextOnEdges: boolean;
  sidebarVisible: boolean;
  navbarVisible: boolean;
  mainColor: availableColorsType;
  darkModeActive: boolean;
  changeMainColor: (color: availableColorsType) => void;
  toggleLocationSidebar: () => void;
  toggleSidebarVisibility: () => void;
  toggleNavbarVisibility: () => void;
  handleSidebarWidthChange: (e: React.DragEvent<HTMLDivElement>) => void;
  toggleTextMode: () => void;
  toggleDarkMode: () => void;
}

export const checkLocalStorage = () => {

  if(!localStorage.locationSidebar){
    const locationSidebar = LocationSidebar.Left
    localStorage.setItem("locationSidebar", JSON.stringify(locationSidebar))
  }

  if(!localStorage.widthSidebar){
    const widthSidebar = Math.floor(window.innerWidth * 0.2)
    localStorage.setItem("widthSidebar", JSON.stringify(widthSidebar))
  }

  if(!localStorage.mainColor){
    localStorage.setItem("mainColor", JSON.stringify(availableColors[0]))
  }

  if(!localStorage.darkModeActive){
    localStorage.setItem("darkModeActive", JSON.stringify(false))
  }else{
    if (JSON.parse(localStorage.darkModeActive as string)){
      const rootElement = document.querySelector("div#root")!
      rootElement.classList.add("dark");
    }
  }
}

checkLocalStorage();

export const useCustomizationStore = create(devtools<CustomizationStoreState>((set, get) => ({
  locationSidebar: JSON.parse(localStorage.locationSidebar as string) as LocationSidebar,
  widthSidebar: JSON.parse(localStorage.widthSidebar as string) as number,
  showTextOnEdges: false,
  sidebarVisible: true,
  navbarVisible: true,
  mainColor: JSON.parse(localStorage.mainColor as string) as availableColorsType,
  darkModeActive: JSON.parse(localStorage.darkModeActive as string) as boolean, 
  changeMainColor: (color: availableColorsType) => {
    setColorVariants(color)

    set({
      mainColor: color,
    })
  },
  toggleLocationSidebar: () => {
    let location = get().locationSidebar
    
    if (location === LocationSidebar.Left){
      location = LocationSidebar.Right
    } else {
      location = LocationSidebar.Left
    }

    localStorage.setItem("locationSidebar", JSON.stringify( location) )  

    set({
      locationSidebar: location,
    })
  },
  toggleSidebarVisibility:() => {
    set({
      sidebarVisible: !get().sidebarVisible,
    })
  },
  toggleNavbarVisibility:() => {
    set({
      navbarVisible: !get().navbarVisible,
    })
  },
  handleSidebarWidthChange: (e: React.DragEvent<HTMLDivElement>) => {
    let widthSidebar

    if (get().locationSidebar === LocationSidebar.Left){
      widthSidebar = e.clientX; 
    } else {
      widthSidebar = window.innerWidth - e.clientX; 
    }

    localStorage.setItem("widthSidebar", JSON.stringify(widthSidebar) );
    set({
      widthSidebar: widthSidebar,
      sidebarVisible: true,
    })
  },
  toggleTextMode: () => {
    set({
        showTextOnEdges: !get().showTextOnEdges
    })
  },
  toggleDarkMode: () => {
    const rootElement = document.querySelector("div#root")!
    rootElement.classList.toggle("dark");

    localStorage.setItem("darkModeActive", JSON.stringify(!get().darkModeActive))

    set({
      darkModeActive: !get().darkModeActive
    })
  }
})));