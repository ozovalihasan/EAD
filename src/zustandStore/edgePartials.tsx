export type HasOneEdgePartialType =  {
  type: "hasOne",
  label: "has one"
}

export const hasOneEdgePartial: HasOneEdgePartialType= {
  type: "hasOne",
  label: "has one"
}

export type HasManyEdgePartialType = {
  type: "hasMany",
  label: "has many"
}

export const hasManyEdgePartial: HasManyEdgePartialType = {
  type: "hasMany",
  label: "has many"
}

export type ThroughEdgePartialType = {
  type: "through",
  label: "through"
}

export const throughEdgePartial: ThroughEdgePartialType = {
  type: "through",
  label: "through"
}