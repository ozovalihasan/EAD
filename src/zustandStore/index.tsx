export * from './edgePartials';
export * from './nodePartials';
export * from './store';
