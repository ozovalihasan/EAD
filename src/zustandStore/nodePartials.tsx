export const entityNodePartial: {
  type: "entity"
} = {
  type: "entity"
}