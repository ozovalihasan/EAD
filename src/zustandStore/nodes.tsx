import { Node } from "react-flow-renderer";

const initialNodes: Node[] = [
  {
    "id": "4",
    "type": "entity",
    "position": {
        "x": 465,
        "y": 150
    },
    "data": {
        "tableId": "1",
        "name": "Physician"
    },
    "width": 138,
    "height": 45
  },
  {
      "id": "5",
      "type": "entity",
      "position": {
          "x": 585,
          "y": 255
      },
      "data": {
          "tableId": "2",
          "name": "Appointment"
      },
      "width": 138,
      "height": 45,
      "selected": false,
      "positionAbsolute": {
          "x": 585,
          "y": 255
      },
      "dragging": false
  },
  {
      "id": "6",
      "type": "entity",
      "position": {
          "x": 720,
          "y": 150
      },
      "data": {
          "tableId": "3",
          "name": "Patient"
      },
      "width": 138,
      "height": 45,
      "selected": false,
      "positionAbsolute": {
          "x": 720,
          "y": 150
      },
      "dragging": false
  }
];

export default initialNodes