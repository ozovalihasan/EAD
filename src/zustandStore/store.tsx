import create from 'zustand';
import {
  Connection,
  Edge,
  EdgeChange,
  Node,
  NodeChange,
  OnNodesChange,
  OnEdgesChange,
  OnConnect,
  applyNodeChanges,
  applyEdgeChanges,
  NodeRemoveChange,
} from 'react-flow-renderer';

import { devtools } from 'zustand/middleware'
import produce from "immer"
import initialTables, { TablesType } from './tables';
import initialNodes from './nodes';
import initialEdges from './edges';
import { EntityNodeType } from '@/components';
import transform_0_4_x_to_0_4_5 from './helpers/transform_0_4_x_to_0_4_5';

export const initialIdCounter = (initialTables: TablesType, initialNodes: Node[], initialEdges: Edge[]): number => {
  
  const tableIds = Object.keys(initialTables).map(( tableId )=> parseInt(tableId ))
  const nodeIds = initialNodes.map((node)=> parseInt(node.id) )
  const edgeIds = initialEdges.map((edge)=> parseInt(edge.id) )

  const max = Math.max( ...tableIds.concat(nodeIds).concat(edgeIds).concat([0]) ) 
  
  return (max + 1)
}

export interface ThroughEdgeDataType {
  throughNodeId: string
}

export type HasAnyEdgeType = Omit<Edge, "data"> & {
  type: "hasMany" | "hasOne"
};

export type ThroughEdgeType = Omit<Edge, "data"> &{
  data: ThroughEdgeDataType,
  type: "through"
}

export type CustomEdgeType = HasAnyEdgeType | ThroughEdgeType;

export interface State {
  version: string;
  idCounter: number;
  nodes: EntityNodeType[];
  edges: CustomEdgeType[];
  tables: TablesType;
  connectionStartNodeId: string | null;
  isConnectContinue: boolean;
  isMouseOnNode: boolean;
  isMouseOnEdge: boolean;
  selectedNodeIdForThrough: string | null;
  mouseOnEdgeId: string | null;
  mouseOnNodeId: string | null;
  associationType: string;
  onNodeMouseEnter: (_: React.MouseEvent, node: Node) => void; 
  onEdgeMouseEnter: (_: React.MouseEvent, edge: Edge) => void; 
  onNodesChange: OnNodesChange;
  onEdgesChange: OnEdgesChange;
  onConnect: OnConnect;
  onMouseEnterThrough: (nodeId: string) => void;
  onChangeAssociationType: (associationType: string, id: string) => void;
  onTableNameChange: (event: React.ChangeEvent<HTMLInputElement>, tableId: string) => void;
  onAttributeNameChange: (event: React.ChangeEvent<HTMLInputElement>, tableId: string, attributeId: string) => void;
  onAttributeTypeChange: (event: React.ChangeEvent<HTMLSelectElement>, tableId: string, attributeId: string) => void;
  addNode: (node: EntityNodeType) => void;
  changeTableSuperClass : (event: React.ChangeEvent<HTMLSelectElement>, tableId: string) => void;
  addTable: () => void;
  addAttribute: (tableId: string ) => void;
  removeAttribute: (tableId: string, attributeId: string ) => void;
  removeTable: (tableId: string ) => void;
  increaseIdCounter: () => void;
  onConnectStart: () => void;
  onConnectEnd: () => void;
  onNodeMouseLeave: () => void;
  onEdgeMouseLeave: () => void;
  onNodeTableChange: (event: React.ChangeEvent<HTMLSelectElement>, nodeId: string) => void;
  onNodeInputChange: (event: React.ChangeEvent<HTMLInputElement>, nodeId: string) => void;
  resetStore: () => void;
  uploadStore: (event: React.ChangeEvent<HTMLInputElement>) => void;
}

const useStore = create(devtools<State>((set, get) => ({
    version: "0.4.5",  
    idCounter: initialIdCounter(initialTables, initialNodes, initialEdges) ,
    associationType: "has_one",
    nodes: initialNodes,
    edges: initialEdges,
    tables: initialTables,
    connectionStartNodeId: null,
    isConnectContinue: false,
    isMouseOnNode: false,
    isMouseOnEdge: false,
    mouseOnNodeId: null,
    mouseOnEdgeId: null,
    selectedNodeIdForThrough: null,
    onConnectStart: (() => {
      set({
        isConnectContinue: true
      })
    }),
    onConnectEnd: (() => {
      set({
        isConnectContinue: false,
        selectedNodeIdForThrough: null
      })
    }),
    onEdgeMouseEnter: ((_: React.MouseEvent, edge: Edge) => {
      if (!get().isConnectContinue){
        set({
          isMouseOnEdge: true,
          mouseOnEdgeId: edge.id
        })
      }
    }),
    onEdgeMouseLeave: (() => {
      if (!get().isConnectContinue){
        set({
          isMouseOnEdge: false,
          mouseOnEdgeId: null
        })
      }
    }),
    onNodeMouseEnter: ((_: React.MouseEvent, node: Node) => {
      set({
        isMouseOnNode: true,
        mouseOnNodeId: node.id
      })
    }),
    onNodeMouseLeave: (() => {
      set({
        isMouseOnNode: false,
        mouseOnNodeId: null
      })  
    }),
    onTableNameChange: ((event: React.ChangeEvent<HTMLInputElement>, tableId: string) => {
      set(
        produce((state: State) => {
          state.tables[tableId].name = event.target.value
        })
      );
    }),
    onAttributeNameChange: ((event: React.ChangeEvent<HTMLInputElement>, tableId: string, attributeId: string) => {
      set(
        produce((state: State) => {
          state.tables[tableId].attributes[attributeId].name = event.target.value
        })
      );
    }),
    onAttributeTypeChange: ((event: React.ChangeEvent<HTMLSelectElement>, tableId: string, attributeId: string) => {
      set(
        produce((state: State) => {
          state.tables[tableId].attributes[attributeId].type = event.target.value
        })
      );
    }),
    addTable: (() => {
      set(produce((state: State) => {
        state.tables[get().idCounter.toString()] = {name: "", attributes: {}, superclassId: ""},
        state.idCounter ++
      }))
    }),
    changeTableSuperClass : ((event: React.ChangeEvent<HTMLSelectElement>, tableId: string) => {
      set(
        produce((state: State) => {
          state.tables[tableId].superclassId = event.target.value
        })
      )
    }),
    addAttribute: ((tableId: string ) => {
      set(produce((state: State) => {
        state.tables[tableId].attributes[get().idCounter.toString()] = {name: "", type: "string"},
        state.idCounter ++
      }))
    }),
    removeAttribute: ((tableId: string, attributeId: string ) => {
      set(produce((state: State) => {
        delete state.tables[tableId].attributes[attributeId]
      }))
    }),
    removeTable: ((tableId: string ) => {
      set(produce((state: State) => {
        delete state.tables[tableId]
        state.nodes = state.nodes.filter((node) => {
          state.edges = state.edges.filter((edge: CustomEdgeType) => (

            node.data.tableId !== tableId || 
            (
              (edge.source !== node.id) && 
              (edge.target !== node.id) && 
              (edge.type !== "through" || edge.data.throughNodeId !== node.id)
            )
            
          ))
          return (node.data.tableId !== tableId)
        })
        
        Object.values(state.tables).forEach(table => {
          if (table.superclassId === tableId){
            table.superclassId = ""
          }
        })
      }))
    }),
    onNodeTableChange: (
      (event: React.ChangeEvent<HTMLSelectElement>, nodeId: string) => {
        set(produce((state: State) => {
          const node: EntityNodeType = (state.nodes.find(node => node.id === nodeId))!
          node.data.tableId = event.target.value
        }))
      }

    ),
    increaseIdCounter: (() =>{
      set({
          idCounter: get().idCounter + 1
      })
    }),
    onNodeInputChange: (event: React.ChangeEvent<HTMLInputElement>, nodeId: string) =>{
      
      set(produce((state: State) => {
        const node: EntityNodeType = (state.nodes.find(node => node.id === nodeId))!
        node.data.name = event.target.value
      }))
    },
    addNode: (node: EntityNodeType) =>{
      set({
          nodes: get().nodes.concat(node),
      })
    },
    onNodesChange: (changes: NodeChange[]) => {
      let edges = get().edges
      if (changes[0].type === "remove"){
        edges = get().edges.filter((edge) => edge.type !== 'through' || edge.data.throughNodeId !== (changes[0] as NodeRemoveChange).id)
      }
      
      set({
        nodes: applyNodeChanges(changes, get().nodes) as EntityNodeType[],
        edges: edges,
      });
    },
    onEdgesChange: (changes: EdgeChange[]) => {
      
      set({
        edges: applyEdgeChanges(changes, get().edges) as CustomEdgeType[],
      });
    },
    onMouseEnterThrough: (nodeId: string) => {
      set({
        selectedNodeIdForThrough: nodeId
      })
    }, 
    onConnect: (connection: Connection) => {
      const id = get().idCounter
      const edgeBase: Omit<CustomEdgeType, "label" | "type"> = {       
        id: id.toString(), 
        source: connection.source!, 
        target: connection.target!, 
        sourceHandle: connection.sourceHandle,
        targetHandle: connection.targetHandle,
      }
   
      let edge: CustomEdgeType;
      
      if (get().associationType === "has_one"){
        
        edge = { 
          ...edgeBase,
          label: 'has one', 
          type: "hasOne",
        }
       } else if (get().associationType === "has_many") {
        edge = { 
          ...edgeBase,
          label: 'has many',
          type: "hasMany",
        }
      } else {
        const selectedNodeIdForThrough = get().selectedNodeIdForThrough
        if (selectedNodeIdForThrough === null){
          alert("It is necessary to select a node to define through association.");
          return 
        }
        
        edge = { 
          ...edgeBase,
          label: 'through',
          type: "through",
          data: {throughNodeId: get().selectedNodeIdForThrough! },
        }
      }

      set({
        edges: get().edges.concat(edge),
        idCounter: get().idCounter + 1,
        selectedNodeIdForThrough: null,
      })
    },
   
    onChangeAssociationType: (associationType: string, id: string) =>{
      set({
        associationType: associationType,
        connectionStartNodeId: id
      });
    },
    resetStore: () => {
      if (window.confirm("EAD will be reset permanently?")){

        set({
          idCounter: 1,
          tables: {},
          nodes: [],
          edges: [],
        })
      }

    },
    uploadStore: (event: React.ChangeEvent<HTMLInputElement>) => {
      const fileReader = new FileReader();
      fileReader.onload = (event) => {
        let data: State;
        if (event.target && (typeof event.target.result === 'string')){
          data = JSON.parse(event.target.result) as State;
          
          if (["0.4.0", "0.4.1", "0.4.2", "0.4.3", "0.4.4", "0.4.5"].includes(data.version)) {
            
            set(
              transform_0_4_x_to_0_4_5(data)
            ) 
          
          } else {
            alert(`The version of your file is v${data.version}. It is not compatible with the version used(v0.4.5).`);  
          }
          
        }else{
          alert("An invalid file is installed. Please check your file.");
        }
        
      
      };
  
      fileReader.readAsText((event.target.files!)[0], 'UTF-8');
    }
    
  })
));

export default useStore;
