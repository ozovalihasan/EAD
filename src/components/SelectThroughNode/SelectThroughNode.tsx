import { useStore } from '@/zustandStore';

export const SelectThroughNode = ({nodeId} : {nodeId: string}) => {

  const onMouseEnterThrough = useStore(store => store.onMouseEnterThrough)

  const visibleThroughSelectArea = useStore(store =>  (
    store.isConnectContinue && 
    store.associationType === "through" && 
    ( store.connectionStartNodeId !== nodeId )
  ))
  
  return (
    <div 
      className={`
        text-first-100 bg-first-500 w-16 h-full flex 
        absolute right-full bottom-0 rounded-l-md text-tiny  
        ${visibleThroughSelectArea ? "visible" : "hidden"}
      `} 
      
      onMouseEnter={() => onMouseEnterThrough(nodeId)}
    >
      <div className="m-auto">
        through
      </div>
    </div>
  )
}