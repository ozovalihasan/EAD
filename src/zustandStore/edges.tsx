import { CustomEdgeType } from "@/zustandStore/store";
import { hasManyEdgePartial, hasOneEdgePartial, throughEdgePartial } from "@/zustandStore/edgePartials";

const initialEdges: CustomEdgeType[] = [
  {
    "id": "7",
    "source": "4",
    "target": "5",
    "sourceHandle": "bottom2",
    "targetHandle": "top",
    "selected": false,
    ...hasManyEdgePartial,
    data: {
      optional: true
    }
  },
  {
    "id": "8",
    "source": "6",
    "target": "5",
    "sourceHandle": "bottom1",
    "targetHandle": "top",
    "selected": false,
    ...hasOneEdgePartial,
    data: {
      optional: false
    }
  },
  {
    "id": "9",
    "source": "4",
    "target": "6",
    "sourceHandle": "bottom3",
    "targetHandle": "top",
    "selected": false,
    ...throughEdgePartial,
    "data": {
        "throughNodeId": "5"
    }
  }  
];

export default initialEdges
