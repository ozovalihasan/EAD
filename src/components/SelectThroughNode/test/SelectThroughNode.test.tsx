import { SelectThroughNode } from '../SelectThroughNode';
import { fireEvent, render, renderHook, screen } from "@testing-library/react";
import { useStore } from '@/zustandStore';

let renderReadyComponent: JSX.Element;

beforeEach(() => {
  
  useStore.setState({ 
    isConnectContinue: false,
    associationType: "has_many",
    connectionStartNodeId: "111",
    onMouseEnterThrough: jest.fn(), 
  })
  
  renderReadyComponent = (
    <SelectThroughNode 
      nodeId="111"
    />
  );
});

describe('<SelectThroughNode />', () => {
  
  describe('if it is available to be visible', () => {
    beforeEach(() => {
  
      useStore.setState({ 
        isConnectContinue: true,
        associationType: "through",
        connectionStartNodeId: "112",
      })
  
    });

    it('has a "visible" class', () => {
      
      render(renderReadyComponent );

      const mainElement = screen.getByText(/through/i).parentNode as HTMLElement

      expect(mainElement.classList).toContain("visible");
  
    });
  
    it('calls a function when mouse enters', () => {
      render(renderReadyComponent );
      const { result } = renderHook(() => useStore());
      
      const mainElement = screen.getByText(/through/i)
  
      expect(result.current.onMouseEnterThrough).toHaveBeenCalledTimes(0);

      fireEvent.mouseEnter(mainElement);

      expect(result.current.onMouseEnterThrough).toHaveBeenCalledTimes(1);
  
    });
    
  
    it('renders correctly', () => {
      const renderedContainer = render(renderReadyComponent );
      expect(renderedContainer).toMatchSnapshot();
    });

  })
  
  describe('if it is not available to be visible', () => {

    it('has a "hidden" class', () => {
      render(renderReadyComponent );

      const mainElement = screen.getByText(/through/i).parentNode as HTMLElement

      expect(mainElement.classList).toContain("hidden");
  
    });
  
    it('renders correctly', () => {
      const renderedContainer = render(renderReadyComponent );
      expect(renderedContainer).toMatchSnapshot();
    });

  })
});