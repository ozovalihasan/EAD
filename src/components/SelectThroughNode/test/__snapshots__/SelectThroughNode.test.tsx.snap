// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SelectThroughNode /> if it is available to be visible renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="
        text-first-100 bg-first-500 w-16 h-full flex 
        absolute right-full bottom-0 rounded-l-md text-tiny  
        visible
      "
      >
        <div
          class="m-auto"
        >
          through
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="
        text-first-100 bg-first-500 w-16 h-full flex 
        absolute right-full bottom-0 rounded-l-md text-tiny  
        visible
      "
    >
      <div
        class="m-auto"
      >
        through
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<SelectThroughNode /> if it is not available to be visible renders correctly 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="
        text-first-100 bg-first-500 w-16 h-full flex 
        absolute right-full bottom-0 rounded-l-md text-tiny  
        hidden
      "
      >
        <div
          class="m-auto"
        >
          through
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="
        text-first-100 bg-first-500 w-16 h-full flex 
        absolute right-full bottom-0 rounded-l-md text-tiny  
        hidden
      "
    >
      <div
        class="m-auto"
      >
        through
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
